File: Adda247Assignment/app/src/main/java/com/demo/myapplication/MainActivity.java
package com.demo.myapplication;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import com.demo.myapplication.Adapter.CustomAdapter;
import com.demo.myapplication.Model.ApiResponce;
import com.demo.myapplication.Model.UserData;
import com.demo.myapplication.ViewModel.PostViewModel;
import com.demo.myapplication.databinding.ActivityMainBinding;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements LifecycleOwner {
    ActivityMainBinding activityMainBinding;
    PostViewModel postViewModel;
    CustomAdapter customAdapter;
    RecyclerView recyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityMainBinding = ActivityMainBinding.inflate(getLayoutInflater());
        View view = activityMainBinding.getRoot();
        setContentView(view);
        recyclerView = activityMainBinding.recycle;

        initUi();
        postViewModel= ViewModelProviders.of(this).get(PostViewModel.class);
        postViewModel.getPostData().observe(this, new Observer<ApiResponce>() {
            @Override
            public void onChanged(ApiResponce apiResponce) {
                Log.d("test", apiResponce.toString());
                getAllData(apiResponce.getData());


            }


        });



    }

    private void getAllData(List<UserData> data) {
        customAdapter.refreshList(data);
    }

    private void initUi() {
        customAdapter = new CustomAdapter(new ArrayList<UserData>());
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(customAdapter);
    }
}

File: KotlinAssignmentAdda247/app/src/main/java/com/demo/kotlinassignmentadda247/MainActivity.kt
package com.demo.kotlinassignmentadda247

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.demo.kotlinassignmentadda247.Adapter.CustomAdapter
import com.demo.kotlinassignmentadda247.Network.RetrofitService
import com.demo.kotlinassignmentadda247.Repository.ApiRepo
import com.demo.kotlinassignmentadda247.ViewModel.PostViewModel
import com.demo.kotlinassignmentadda247.ViewModel.PostViewModelFactory
import com.demo.kotlinassignmentadda247.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private val TAG = "MainActivity"
    private lateinit var binding: ActivityMainBinding

    lateinit var viewModel: PostViewModel

    private val retrofitService = RetrofitService.getInstance()
    val adapter = CustomAdapter()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
      val layoutManager:  RecyclerView.LayoutManager = LinearLayoutManager(this)

        viewModel = ViewModelProvider(this, PostViewModelFactory(ApiRepo(retrofitService)))
            .get(PostViewModel::class.java)
        binding.recycle.layoutManager= layoutManager

        binding.recycle.adapter = adapter

        viewModel.userList.observe(this, Observer {
            Log.d(TAG, "onCreate: $it")
            adapter.refreshList(it)
        })

        viewModel.errorMessage.observe(this, Observer {
            Log.d(TAG, "onCreate: $it")
        })

        viewModel.getData()
    }
}